drivers: gpio: move non-standard dts flags to be soc specific

henrikbrixandersen opened the pull request:
Reserve the upper 8 bits of `gpio_dt_flags_t` for SoC specific flags and move the non-standard, hardware-specific GPIO devicetree flags (IO voltage level, drive strength, debounce filter) from the generic `dt-bindings/gpio/gpio.h` header to SoC specific dt-bindings headers.
    
Some of the SoC specific dt-bindings flags take up more bits than necessary in order to retain backwards compatibility with the deprecated GPIO flags. The width of these fields can be reduced/optimized once the deprecated flags are removed.
    
Remove hardcoded use of `GPIO_INT_DEBOUNCE` in GPIO client drivers. This flag can now be set in the devicetree for boards/SoCs with debounce filter support. The SoC specific debounce flags have had the `_INT` part of their name removed since these flag must be passed to `gpio_pin_configure()`, not `gpio_pin_interrupt_configure()`.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: gpio: move non-standard dts flags to be soc specific

Reserve the upper 8 bits of gpio_dt_flags_t for SoC specific flags and
move the non-standard, hardware-specific GPIO devicetree flags (IO
voltage level, drive strength, debounce filter) from the generic
dt-bindings/gpio/gpio.h header to SoC specific dt-bindings headers.

Some of the SoC specific dt-bindings flags take up more bits than
necessary in order to retain backwards compatibility with the deprecated
GPIO flags. The width of these fields can be reduced/optimized once the
deprecated flags are removed.

Remove hardcoded use of GPIO_INT_DEBOUNCE in GPIO client drivers. This
flag can now be set in the devicetree for boards/SoCs with debounce
filter support. The SoC specific debounce flags have had the _INT part
of their name removed since these flag must be passed to
gpio_pin_configure(), not gpio_pin_interrupt_configure().

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

keith-zephyr commented on include/dt-bindings/gpio/gpio.h:
```diff
@@ -102,6 +102,59 @@
+ * @{
+ */
+/** @cond INTERNAL_HIDDEN */
+#define GPIO_DS_LOW_POS 20
```
gpio_dt_flags_t is currently defined as uint8_t.  gpio_dt_flags_t should be updated as part of this change.  

If the GPIO flags in devicetree are going to be extended past 8 bits, It probably makes sense to just eliminate gpio_dt_flags_t and only use gpio_flags_t throughout Zephyr.

henrikbrixandersen replied:
I have opted to extend `gpio_dt_flags_t` to 16 bits, but kept it for now. I agree that we should consider removing it, but I would like to keep that out of this PR.

keith-zephyr replied:
We (Google) have a use case for specifying the GPIO_INPUT/OUTPUT and the GPIO_OUTPUT_INIT_LOW/INIT_HIGH flags in the devicetree. So I'm in favor of extended gpio_dt_flags_t to 32 bits and making all the flags available in the bindings header.

henrikbrixandersen replied:
That would make all of our GPIO flags Zephyr-specific devicetree extensions. I am not sure, that's a good plan.

henrikbrixandersen replied:
I will bring this up on todays dev-review meeting.

henrikbrixandersen replied:
We discussed this a bit on todays API meeting, but didn't reach a conclusion.
@mbolivar-nordic Any input on this?

mbolivar-nordic replied:
> @mbolivar-nordic Any input on this?

I'd like to get @mnkp to approve this. It's always been his design decision as GPIO maintainer that the Zephyr devicetree GPIO flags match Linux's exactly, so I think this is his decision.

carlescufi replied:
@henrikbrixandersen could you please comment on the fact that we'd diverge from the Linux GPIO flags in DT?

henrikbrixandersen replied:
@carlescufi That's already present: `The drive strength flags are a Zephyr specific extension ...`

keith-zephyr reviewed (dismissed):
If the dt-bindings are going to be extended, I think all the GPIO flags should be moved into the dt-bindings instead of just the drive strength flags.

galak commented on include/dt-bindings/gpio/gpio.h:
```diff
@@ -102,6 +102,59 @@
 
 /** @} */
 
+/**
```
Add comment about these being zephyr specific like the voltages are.

henrikbrixandersen replied:
Added, thanks.

carlescufi commented:
@henrikbrixandersen ping

keith-zephyr commented:
I'd like to see this change get approved.  What's the current status of this PR?

henrikbrixandersen commented:
> I'd like to see this change get approved. What's the current status of this PR?

@keith-zephyr Thanks for reminding me. I have updated this PR to extend the size of `gpio_dt_flags_t` from 8 to 16 bits and moved the flags accordingly.

mnkp commented:
In principle we want to follow [Linux gpio bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt) and they don't define a property to set the drive strength in DTS. The drive strength property is defined only in [Linux pinctrl bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml) as:
```

  drive-strength:
    $ref: /schemas/types.yaml#/definitions/uint32
    description: sink or source at most X mA
```
It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_`  flags don't match well hardware capabilities of devices supported by Zephyr.

gmarull commented:
> In principle we want to follow [Linux gpio bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt?rgh-link-date=2022-02-07T03%3A20%3A32Z) and they don't define a property to set the drive strength in DTS. The drive strength property is defined only in [Linux pinctrl bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml?rgh-link-date=2022-02-07T03%3A20%3A32Z) as:
> 
> ```
> 
>   drive-strength:
>     $ref: /schemas/types.yaml#/definitions/uint32
>     description: sink or source at most X mA
> ```
> 
> It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_` flags don't match well hardware capabilities of devices supported by Zephyr.

The Zephyr GPIO flags do not have units:

```
 * `DFLT` - The lowest drive strength supported by the HW
 * `ALT` - The highest drive strength supported by the HW
```

They really allow setting the lowest/highest value, which seems portable enough and useful for simple cases like e.g., driving LEDs. Drive strength is one of those vendor-specific stuff that can hardly be generalized.

henrikbrixandersen commented:
> It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_` flags don't match well hardware capabilities of devices supported by Zephyr.

This doesn't align with the decision of not using pinctrl for GPIO in Zephyr. This was discussed on the API meeting 2021-11-02 (see https://github.com/zephyrproject-rtos/zephyr/pull/39767#issuecomment-957902226).

carlescufi commented:
API meeting

> * Concern as to there is duplication between pinctrl standard properties and the macros in this Pull Request
> * The pinctrl node is not used for pure GPIOs, hence there is no duplication possible. This is already documented here: https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-control-vs-gpio

- Counterproposal from @mnkp: Reserve the upper 8 bits of `gpio_dt_flags_t` so that they are implementation-dependent. This would be similar to `pinctrl_soc_pin_t` but only for the upper 8-bit of the 16-bit flags. Also remove the current API drive strength and debounce flags.

mbolivar-nordic commented on samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c:
```diff
@@ -64,8 +64,7 @@ void app_gpio_init(void)
 
 	button_device[0] = device_get_binding(DT_GPIO_LABEL(SW0_NODE, gpios));
 	gpio_pin_configure(button_device[0], DT_GPIO_PIN(SW0_NODE, gpios),
-			   GPIO_INPUT | GPIO_INT_DEBOUNCE |
```
Removing the debounce flag here seems to change the expected behavior. The sample documents that it works with nrf52840dk_nrf52840, so it seems like this should at least add an overlay for that board specifying the right flags there, no? I have the same question for other similar changes elsewhere in this patch.

mbolivar-nordic replied:
Or maybe all the relevant board DTS files need to have the debounce flags added to their button nodes?

henrikbrixandersen replied:
@mbolivar-nordic That's the thing, really. The sample support `nrf52840dk_nrf52840`, but the nRF GPIO driver does not support `GPIO_INT_DEBOUNCE` at all, so this flag has no effect.

One could argue that all button nodes for boards with a SoC supporting GPIO debouncing should have that flag added, but this is no simple task (60+ boards from what I count) - and some of these may be doing hardware button debouncing, rendering this flag unnecessarily or even counterproductive.

As for the other drivers where I have removed this flag, they are drivers for external LoRa chips where the interrupt lines where configured with debouncing enabled. This should not be needed at all.

mbolivar-nordic replied:
OK, fair enoguh

henrikbrixandersen commented on include/dt-bindings/gpio/andestech-atcgpio100.h:
```diff
@@ -0,0 +1,18 @@
+ * specified by the Linux GPIO binding. Only applicable for Andestech Technology
+ * Corporation ATCGPIO100 SoCs.
+ */
+#define GPIO_INT_DEBOUNCE (1U << 8)
```
@gmarull I have updated this PR to namespace the SoC specific GPIO DTS flags.

gmarull reviewed (dismissed):
LGTM. In some cases, like nRF, we can now offer more meaningful options for the users. Will look at it after this gets in.

mbolivar-nordic reviewed (dismissed):
We need an index of device drivers in the documentation that includes information on driver-specific features more than ever with changes like this going in...

mnkp commented on include/drivers/gpio.h:
```diff
@@ -201,79 +201,6 @@ extern "C" {
- * @note Drivers that do not support a debounce feature should ignore
- * this flag rather than rejecting the configuration with -ENOTSUP.
- */
-#define GPIO_INT_DEBOUNCE              (1U << 19)
```
We cannot remove the macros, need to deprecate them first, e.g.:
```
#define GPIO_INT_DEBOUNCE              __DEPRECATED_MACRO (1U << 8)
```
The same holds true for drive strength flags.

henrikbrixandersen replied:
It's tricky to both deprecate them and make them SoC specific without overlapping.

mnkp replied:
When going the `__DEPRECATED_MACRO` way we would need to make sure that the code compiles and the drivers work just as before. The only difference being the warning message generated by the compiler.

This is possible as long as we use the same bits in the SoC specific flags. I.e. in `include/drivers/gpio.h`
```
#define GPIO_DS_LOW_POS __DEPRECATED_MACRO 9
#define GPIO_DS_HIGH_POS __DEPRECATED_MACRO 11
```
and then in the SoC specific flags e.g. `include/dt-bindings/gpio/nordic-nrf-gpio.h`
```
#define NRF_GPIO_DS_LOW_POS 9
#define NRF_GPIO_DS_HIGH_POS 11
```
We need to keep the deprecated macros for two releases. When they are gone the SoC specific flags can be cleaned up, if required.

henrikbrixandersen replied:
@mnkp Thank you for the feedback. I modified this PR to do that.

mnkp reviewed (changes requested):
Thanks @henrikbrixandersen, great work! I have some minor but blocking comments. Also, I think we should squash "drivers: gpio: allow specifying GPIO drive strength/debounce flags in dts" commit. I.e. implement all the changes in a single commit.

henrikbrixandersen commented:
> Thanks @henrikbrixandersen, great work! I have some minor but blocking comments. Also, I think we should squash "drivers: gpio: allow specifying GPIO drive strength/debounce flags in dts" commit. I.e. implement all the changes in a single commit.

@mnkp Thank you for reviewing. I have squashed to two commits into one. Please see my question on how you'd like to handle deprecation of the old macros.

henrikbrixandersen commented:
@mnkp Ping?

anangl commented on drivers/gpio/gpio_andes_atcgpio100.c:
```diff
@@ -140,7 +141,7 @@ static int gpio_atcgpio100_config(const struct device *port,
 
 		/* Set de-bounce */
-		if (flags & GPIO_INT_DEBOUNCE) {
+		if (flags & ATCGPIO100_GPIO_INT_DEBOUNCE) {
```
I think we should get rid of this misleading `INT_` component from the names of debounce flags. Unlike all other `GPIO_INT_*` flags, the debounce flag is supposed to be used in calls to `gpio_pin_configure()`, not `gpio_pin_interrupt_configure()`.

henrikbrixandersen replied:
@anangl Thank you for reviewing. I have removed the `_INT` part of the name for the SoC specific debounce flags.

anangl commented on include/dt-bindings/gpio/gpio.h:
```diff
@@ -76,29 +76,26 @@
 
 /** Set pin at the default voltage level */
-#define GPIO_VOLTAGE_DEFAULT	(0U << GPIO_VOLTAGE_POS)
+#define GPIO_VOLTAGE_DEFAULT    (0U << GPIO_VOLTAGE_POS)
```
Shouldn't `__DEPRECATED_MACRO` be added also to `GPIO_VOLTAGE_*` macros?

henrikbrixandersen replied:
As far as I can tell, our tools do not support `__DEPRECATED_MACRO` in dts files.
